passwd_change.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PASSWD_CHANGE 1"
.TH PASSWD_CHANGE 1 "2014-01-15" "3.6" "kadmin-remctl"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
passwd_change \- Change a user's Kerberos password
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpasswd_change\fR \fIuser\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBpasswd_change\fR changes the password for \fIuser\fR, who may be someone
other than the person running the program.  It may only be run by
authorized users (normally Help Desk personnel); all others will receive
an error message.  The person running the program must have a valid ticket
cache when running it and will be asked to reauthenticate as a security
precaution.  Then, \fIuser\fR will be checked against the user database (in
the form of a unified passwd file) and the user's full name will be
displayed for verification if found.  When changing the password for an
account not listed in the passwd file, the user will not be found but one
can elect to continue anyway.  Finally, the user is prompted for the new
password (twice).
.PP
This program uses the remctl protocol to talk to a central server to do
the password change.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
\&\fBpasswd_change\fR needs four configuration parameters: The full path to a
passwd file listing all of the local users (used only to get full names
for verification that one got the right account), the service principal to
use to authenticate to the password changing service, the host on which
the password changing service is running, and the port on which the
service is running.  Defaults for these values are set at compilation
time, but they can be overridden in the \fI/etc/krb5.conf\fR file (or
wherever \fIkrb5.conf\fR path the Kerberos libraries were compiled to use).
.PP
\&\fBpasswd_change\fR looks for a \f(CW\*(C`passwd_change\*(C'\fR section in the [appdefaults]
section of \fIkrb5.conf\fR and supports the following configuration options:
.IP "passwd_file" 4
.IX Item "passwd_file"
The full path to a passwd file listing users.  This file should be
structured like a normal Unix \fI/etc/passwd\fR file; all fields except
username and full name are ignored.  This file is used only to retrieve
the user's full name as a verification step that the correct account's
password is being changed.
.IP "port" 4
.IX Item "port"
The port on which the password changing service is running.  The default
value is 4443.
.IP "server" 4
.IX Item "server"
The host on which the password changing service is running.
.IP "service_principal" 4
.IX Item "service_principal"
The Kerberos service for which to get credentials before changing
passwords.  This should match the Kerberos principal used by the
\&\fBremctld\fR running the password changing service.
.PP
For example, here is the configuration for Stanford:
.PP
.Vb 7
\&    [appdefaults]
\&        passwd_change = {
\&            passwd_file       = /afs/ir/service/etc/passwd.all
\&            server            = password\-change.stanford.edu
\&            port              = 4443
\&            service_principal = service/password\-change@stanford.edu
\&        }
.Ve
.SH "BUGS"
.IX Header "BUGS"
The business of getting the target user's full name from a password file
is an ugly hack that only happens to work at Stanford for legacy reasons.
A saner method should be used, like querying an \s-1LDAP\s0 server.
.SH "AUTHOR"
.IX Header "AUTHOR"
Russ Allbery <eagle@eyrie.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2007, 2008, 2013 The Board of Trustees of the Leland Stanford
Junior University
.PP
Copying and distribution of this file, with or without modification, are
permitted in any medium without royalty provided the copyright notice and
this notice are preserved.  This file is offered as-is, without any
warranty.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIkadmin\-backend\fR\|(8), \fIkpasswd\fR\|(1), \fIpasswd\fR\|(5), \fIremctl\fR\|(1), \fIremctld\fR\|(8)
.PP
This program is part of kadmin-remctl.  The current version is available
from <http://www.eyrie.org/~eagle/software/kadmin\-remctl/>.


NEWS
                    User-Visible kadmin-remctl Changes

kadmin-remctl 3.6 (2014-01-15)

    Add a new per-instance configuration option to set the password
    expiration time for newly-created principals.  Be aware that this only
    controls the initial expiration period.  After the first password
    change, further expiration periods are normally controlled by the KDC
    configuration or policy.

    In the Heimdal backend, map password quality errors on account
    creation or password reset to a generic error.  The kadmin protocol
    doesn't have a mechanism for passing back the rich error message from
    the password quality check, so all failures use the same error string.
    Remap it here, since the error message from Heimdal is of dubious
    accuracy.  This will only apply to sites that have patched Heimdal to
    do password quality checks on administrative operations.

    Update to rra-c-util 5.1:

    * Don't attempt to use Kerberos if no Kerberos error APIs were found.
    * Improve error handling in xasprintf and xvasprintf.
    * Check the return status of snprintf and vsnprintf properly.
    * Preserve errno if snprintf fails in vasprintf replacement.

kadmin-remctl 3.5 (2013-10-10)

    Increase the timeout for initial authentication during a kpasswd
    password change to ten seconds, and the timeout for a successful
    password change to sixty seconds.  The previous timeouts of two
    seconds and thirty seconds was occasionally too short in production.
    Also fix Perl warnings if the initial authentication or password
    change time out.

    In the Heimdal backend, use get instead of list to check whether a
    given principal already exists.  list requires a complete database
    traversal and is much more resource-intensive.

    In the Heimdal backend, recognize the new form of the Heimdal kpasswd
    prompt to repeat the new password.

    Stop mapping password quality errors in the Heimdal kpasswd backend.
    Instead, remove any prefix about an external password quality program
    and pass the rest of the error message back to the user.

    Check the existence of the principal before enabling or disabling it
    in the Heimdal backend so that nonexistent principals report a clearer
    error message instead of an internal error about getAttributes
    failure.

    Update to rra-c-util 4.10:

    * Remove some unnecessary includes.

kadmin-remctl 3.4 (2013-07-09)

    Set the disallow-svr flag on all newly-created principals.  This
    prohibits obtaining service tickets for the principal, which provides
    some hardening against brute force attacks.  Since the create command
    is designed for creation of user principals, not service principals,
    and use of service tickets for user principals is quite obscure and
    rare in Kerberos, this seems like a better default.

    Change the default allowed principal regex to allow two-character user
    principals.  This is just a default and can be overridden by setting
    the allowed key in the configuration.

kadmin-remctl 3.3 (2013-03-25)

    In the Heimdal version of kadmin-backend, retry the kadmin connection
    once if the first connection fails.  This is a workaround for a
    transient networking error that we're seeing at Stanford and therefore
    may not be fully appropriate for other sites, but should hopefully be
    harmless.  Also suppress the standard error output from the Heimdal
    library during connect since Heimdal::Kadm5 does not.

    Clean up error reporting in the Heimdal version of kadmin-backend.
    Use the correct (rather than the documented) way to tell
    Heimdal::Kadm5 to throw exceptions, and ensure that all kadmin
    functions uniformly use the same standard error formatting and exit
    status for kadmin failures.

    Exit with a non-zero status if the check_passwd command rejects the
    password.  Previously, an error would be reported but the backend
    would always report a successful zero status if the password could be
    checked, even if it was rejected.

    The Heimdal version of kadmin-backend now requires the IPC::Run Perl
    module (available from CPAN).

    Produce a better error message when trying to change the password of a
    disabled account with the Heimdal backend.

    When prompting for a username in passwd_change, strip any surrounding
    whitespace from that username before proceeding.

    Update to rra-c-util 4.8:

    * Fix Heimdal libroken probes for old versions of Heimdal.
    * Fix Kerberos header probing with non-standard include paths.
    * Pass --deps to krb5-config if it is supported.
    * Properly find krb5.h on NetBSD systems.
    * Fix stripping of -I/usr/include from krb5-config output.
    * Avoid using krb5-config if specific Kerberos paths are configured.
    * Use PATH_KRB5_CONFIG instead of KRB5_CONFIG to locate krb5-config.
    * Replace concat with xasprintf.
    * xasprintf is now void and always calls the failure handler on error.
    * Improve __attribute__ portability to old GCC or non-GCC compilers.
    * Add -D_FORTIFY_SOURCE=2 to make warnings flags.
    * Probe for ssize_t and replace it in portable/system.h if not found.
    * Include strings.h in portable/system.h if it exists.
    * Add a pointer to rra-c-util in all files.

kadmin-remctl 3.2 (2011-06-09)

    Add support for a separate blacklist of principals whose passwords
    cannot be changed with reset_passwd but who do not themselves have the
    ability to reset passwords.

    Properly handle incorrect password errors from Heimdal's kpasswd.
    Previously, if change_passwd failed because the original password was
    incorrect, kadmin-remctl-heim would output a confusing Expect error.

    When creating principals for Heimdal, set a default set of attributes
    before making the principal.  Heimdal::Kadm5 doesn't seem to be
    working properly with the default attributes from the default

passwd_change.c
/*
 * Allow authorized users to change the passwords of other users.
 *
 * This program allows authorized users to change the passwords of other
 * users.  It talks to a remctl interface via the libremctl library and only
 * implements password changing, with verbose prompting and the ability to
 * read the username of the principal whose password should be changed from
 * the command line).
 *
 * Written by Russ Allbery <eagle@eyrie.org>
 * Copyright 1997, 2007, 2010, 2013, 2014
 *     The Board of Trustees of the Leland Stanford Junior University
 *
 * See LICENSE for licensing terms.
 */

#include <config.h>
#include <portable/krb5.h>
#include <portable/system.h>

#include <ctype.h>
#include <errno.h>
#include <remctl.h>
#include <signal.h>

#include <util/messages-krb5.h>
#include <util/messages.h>
#include <util/xmalloc.h>

/* The full path to the site-wide password file, for real name mapping. */
#ifndef PASSWD_FILE
# define PASSWD_FILE "/full/path/to/passwd/file"
#endif

/* The principal name used for password changing. */
#ifndef PRINCIPAL
# define PRINCIPAL "service/password-change"
#endif

/* The host and port to which the remctl connection should be made. */
#ifndef HOST
# define HOST "password-change.example.org"
#endif
#ifndef PORT
# define PORT 4443
#endif

/* The memory cache used for the password change authentication. */
#define CACHE_NAME "MEMORY:passwd_change"


/*
 * Load a string option from Kerberos appdefaults.
 */
static void
config_string(krb5_context ctx, const char *opt, const char *defval,
              char **result)
{
    krb5_appdefault_string(ctx, "passwd_change", NULL, opt, defval, result);
}


/*
 * Load a number option from Kerberos appdefaults.  The native interface
 * doesn't support numbers, so we actually read a string and then convert.
 */
static void
config_number(krb5_context ctx, const char *opt, int defval, int *result)
{
    char *tmp = NULL;

    krb5_appdefault_string(ctx, "passwd_change", NULL, opt, "", &tmp);
    if (tmp != NULL && tmp[0] != '\0')
        *result = atoi(tmp);
    else
        *result = defval;
    if (tmp != NULL)
        free(tmp);
}


/*
 * Open a connection to kadmind and authenticate to the server.  This creates
 * a new ticket file and obtains the service/password-change ticket which will
 * then be used to change the user's password.  Returns 0 on success and -1 on
 * a failure it's worth retrying.  For some failures, such as memory
 * allocation problems, just dies.
 */
static int
login(krb5_context ctx, char *service)
{
    krb5_error_code status;
    krb5_ccache ccache = NULL;
    krb5_principal princ = NULL;
    krb5_creds creds;
    krb5_get_init_creds_opt *opts;

    /*
     * First of all, we have to figure out what the admin principal is.  We do
     * that by parsing the user's credential cache.
     */
    status = krb5_cc_default(ctx, &ccache);
    if (status != 0) {
        warn_krb5(ctx, status, "cannot open default ticket cache");
        goto fail;
    }
    status = krb5_cc_get_principal(ctx, ccache, &princ);
    if (status != 0) {
        warn_krb5(ctx, status, "cannot get principal name from cache");
        goto fail;
    }
    krb5_cc_close(ctx, ccache);
    ccache = NULL;

    /* Now, we have the user's principal in principal.  Authenticate. */
    status = krb5_get_init_creds_opt_alloc(ctx, &opts);
    if (status != 0)
        die_krb5(ctx, status, "cannot allocate credential options");
    krb5_get_init_creds_opt_set_default_flags(ctx, "passwd_change",
                                              princ->realm, opts);
    memset(&creds, 0, sizeof(creds));
    status = krb5_get_init_creds_password(ctx, &creds, princ, NULL,
                 krb5_prompter_posix, NULL, 0, service, opts);
    if (status != 0) {
        warn_krb5(ctx, status, "authentication failed");
        goto fail;
    }

    /* Put the new credentials into a memory cache. */
    status = krb5_cc_resolve(ctx, CACHE_NAME, &ccache);
    if (status != 0)
        die_krb5(ctx, status, "cannot create memory cache");
    status = krb5_cc_initialize(ctx, ccache, princ);
    if (status != 0)
        die_krb5(ctx, status, "cannot initialize memory cache");
    krb5_free_principal(ctx, princ);
    status = krb5_cc_store_cred(ctx, ccache, &creds);
    if (status != 0)
        die_krb5(ctx, status, "cannot store credentials");
    krb5_cc_close(ctx, ccache);
    krb5_free_cred_contents(ctx, &creds);
    if (putenv((char *) "KRB5CCNAME=" CACHE_NAME) != 0)
        sysdie("putenv of KRB5CCNAME failed");
    return 0;

fail:
    if (ccache != NULL)
        krb5_cc_close(ctx, ccache);
    if (princ != NULL)
        krb5_free_principal(ctx, princ);
    return -1;
}


/*
 * Prompt for a new password and write it into the given pointer.  Returns 0
 * on success, -1 on a retriable failure, and -2 on a permanent failure.
 */
static int
get_password(krb5_context ctx, char **password)
{
    krb5_prompt prompts[2];
    krb5_error_code status;
    int ret = -2;

    /* Set up the prompt structure. */
    prompts[0].prompt = (char *) "New password";
    prompts[0].hidden = 1;
    prompts[0].reply = xcalloc(1, sizeof(*prompts[0].reply));
    prompts[0].reply->data = xmalloc(BUFSIZ);
    prompts[0].reply->length = BUFSIZ;
    prompts[1].prompt = (char *) "Re-enter new password";
    prompts[1].hidden = 1;
    prompts[1].reply = xcalloc(1, sizeof(*prompts[0].reply));
    prompts[1].reply->data = xmalloc(BUFSIZ);
    prompts[1].reply->length = BUFSIZ;

    /* Finally, we can do the actual prompt. */
    status = krb5_prompter_posix(ctx, NULL, NULL, NULL, 2, prompts);
    if (status != 0) {
        warn_krb5(ctx, status, "cannot prompt for a password");
        goto fail;
    }
    if (strcmp(prompts[0].reply->data, prompts[1].reply->data) != 0) {
        warn("passwords don't match");
        ret = -1;
        goto fail;
    }
    *password = prompts[0].reply->data;
    free(prompts[0].reply);
    free(prompts[1].reply->data);
    free(prompts[1].reply);
    return 0;

fail:
    free(prompts[0].reply->data);
    free(prompts[0].reply);
    free(prompts[1].reply->data);
    free(prompts[1].reply);
    return ret;
}


/*
 * Actually change the password of a user.  We prompt for the new password and
 * then call remctl to do the real work.
 */
static int
reset_password(krb5_context ctx, char *principal, const char *service,
               const char *host, unsigned short port)
{
    int status;
    char *password;
    struct remctl_result *result;
    const char *command[5];

    /* Get the new password. */
    do {
        status = get_password(ctx, &password);
        printf("\n");
    } while (status == -1);
    if (status == -2)
        return -1;

    /* Reset the password. */
    command[0] = "password";
    command[1] = "reset";
    command[2] = principal;
    command[3] = password;
    command[4] = NULL;
    result = remctl(host, port, service, command);
    if (result->error != NULL) {
        warn("%s", result->error);
        remctl_result_free(result);
        return -2;
    } else {
        if (result->stderr_len > 0)
            fwrite(result->stderr_buf, result->stderr_len, 1, stderr);
        if (result->stdout_len > 0)
            fwrite(result->stdout_buf, result->stdout_len, 1, stdout);
        if (result->status == 0 && result->stdout_len == 0) {
            printf("Password for %s successfully changed\n", principal);
            return 0;
        } else if (result->status == 2)
            return -2;
        else
            return -1;
    }
}


/*
 * Given a username, find their entry in the site password file and read off
 * their real name.  This is for a double-check verification that one has
 * typed the right account name.  Returns a malloc()d string that the caller
 * is responsible for freeing.  Returns NULL on error or if the username can't
 * be found.
 */
static char *
find_name(char *username, const char *passwd_file)
{
    FILE *passwd;
    char buffer[1024];
    char *name, *search, *start, *end;
    int count;

    /* Build our search string, which is the username followed by a :. */
    xasprintf(&search, "%s:", username);

    /* Open the password file. */
    passwd = fopen(passwd_file, "r");
    if (passwd == NULL) {
        syswarn("unable to open site password file");
        free(search);
        return NULL;
    }

    /*
     * Scan through the password file looking for our search string.  If we
     * find it, grab the fourth field of the password entry, copy it into
     * name, and return it.  Otherwise, return NULL.
     */
    name = NULL;
    do {
        if (!fgets(buffer, sizeof(buffer), passwd))
            break;
        if (!strncmp(buffer, search, strlen(search))) {
            for (start = buffer, count = 0; *start && count < 4; start++)
                if (*start == ':')
                    count++;
            for (end = start + 1; *end && *end != ':'; end++)
                ;
            *end = '\0';
            name = xstrdup(start);
        }
    } while (name == NULL);
    free(search);
    return name;
}


int
main(int argc, char **argv)
{
    krb5_context ctx;
    char *passwd, *service, *host, *p;
    char principal[BUFSIZ], ans[BUFSIZ];
    int port, status, tries;
    char *name;

    /*
     * Set the name of the program, used for error reporting, stripping off
     * the path information first.
     */
    message_program_name = strrchr (argv[0], '/');
    if (message_program_name != NULL)
        message_program_name++;
    else
        message_program_name = argv[0];

    /*
     * Check for a -h or --help flag and spit out a simple usage if one is
     * given, just in case someone tries that.
     */
    if (argc > 1 && (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help"))) {
        printf("Usage: %s [<username>]\n\n", message_program_name);
        printf("Usable by authorized users only, changes the password for "
               "<username>.  The\nusername will be prompted for if not "
               "supplied on the command line.\n");
        exit(0);
    }

    /* Obtain a Kerberos context so that we can look up configuration. */
    status = krb5_init_context(&ctx);
    if (status != 0)
        die_krb5(ctx, status, "cannot initialize Kerberos");
    config_string(ctx, "passwd_file", PASSWD_FILE, &passwd);
    config_string(ctx, "service_principal", PRINCIPAL, &service);
    config_string(ctx, "server", HOST, &host);
    config_number(ctx, "port", PORT, &port);

    /* Authenticate to kadmind. */
    printf("Authenticating to Kerberos....\n");
    if (login(ctx, service))
        exit(1);
    printf("\n");
  
    /*
     * If we were given a username on the command line, use it.  Otherwise,
     * prompt for a username whose password we're changing.  Strip whitespace
     * from the username.
     */
    if (argc > 1)
        strncpy(principal, argv[1], sizeof(principal) - 1);
    else {
        printf("Enter username whose password you wish to change: ");
        if (fgets(principal, sizeof(principal), stdin) == NULL)
            sysdie("error reading username");
        p = principal + strlen(principal) - 1;
        while (p > principal && isspace((unsigned char) *p))
             p--;
        p[1] = '\0';
        for (p = principal; isspace((unsigned char) *p); p++)
            ;
        if (p != principal)
            memmove(principal, p, strlen(p) + 1);
    }

    /* Find the real name and print it out to make sure it's right. */
    name = find_name(principal, passwd);
    if (name == NULL) {
        printf("That username was not found in the password file."
               "  Continue? ");
        if (!fgets(ans, sizeof(ans), stdin) || strncasecmp(ans, "y", 1)) {
            printf("Aborted\n\n");
            free(name);
            exit(1);
        }
    } else {
        printf("%s\t%s\n\nIs this correct? ", principal, name);
        if (!fgets(ans, sizeof(ans), stdin) || strncasecmp(ans, "y", 1)) {
            printf("Aborted\n\n");
            free(name);
            exit(1);
        }
    }

    /* Change the password.  Loop up to five times in the case of an error. */
    for (tries = 0; tries < 5; tries++) {
        status = reset_password(ctx, principal, service, host, port);
        if (!status || status == -2)
            break;
        else
            printf("\n");
    }
    exit(status ? 1 : 0);
}
